orangecontrib/prototypes/widgets/owfactoranalysis.py

```diff
@@ -70,38 +141,47 @@ def __init__(self): # __init__ je konstruktor (prva metoda ki se klice, ko se us
         table.horizontalHeader().hide()
         table.setShowGrid(False)
         box.layout().addWidget(table)
-
+        """
 
     @Inputs.data
     def set_data(self, dataset):
-        self.closeContext()
+        # self.closeContext()
         if dataset is None:
             self.sample = None
         else:
-            self.openContext(dataset.domain)  #Kaj je funkcija contexta
+            # self.openContext(dataset.domain)  # what is the function of context?
+            pass
 
         self.dataset = dataset
-        self.optionsBox.setDisabled(False)
-        self.commit.now() # Takoj poklici metodo commit
+        self.commit.now()
 
     def factor_analysis(self):
-        # Z izbranimi n_componentami izracunaj FA na self.dataset
-        result = FactorAnalysis(self.n_components).fit(self.dataset.X)
-        # Iz spremenljivke result (ki je instance nekega razreda) izlusci samo tabelo ki nas zanima
-        self.components = result.components_
+        # with chosen n_components and depending on the user-selected rotation, calculate the FA on self.dataset
+        if self.setting_for_rotation == 0:
+            result = FactorAnalysis(self.n_components).fit(self.dataset.X)
+        elif self.setting_for_rotation == 1:
+            result = FactorAnalysis(self.n_components, rotation="varimax").fit(self.dataset.X)
+        elif self.setting_for_rotation == 2:
+            result = FactorAnalysis(self.n_components, rotation="quartimax").fit(self.dataset.X)
+        else:
+            print("Error:")
```

Comment:
This has no effect, nobody sees it. In this case, you can assume this won't happen, so you can just change the last `elif` to `else` (well, you don't have to because you don't need `if`-s anyway).

There is however a pattern you can sometimes use for cases where something shouldn't happen and probably doesn't, but you'd like to check it and are OK with widget crashing if it *does* happen (because the widget, for instance, can't continue). In your case, if you'd like to do this, you'd put

`assert self.setting_for_rotation <= 2` before `if`-s. If the condition is not true, the widget will fail with assertion error.

That said, don't do this here, because `setting_for_rotation` will always be correct.

For a good example, `OWPredictions` attribute `self.data`. If `self.data` is not `None`, it constructs ... some model (whatever). Now, a method `_commit_predictions` immediately exits if there is no data:

```
        if not self.data:
            return
```

Later in this function, you can see 

```
datamodel = self.dataview.model()
```

This `datamodel` is the model, which **must** and does and will exist if `self.data` exists. So `datamodel` cannot be `None`. But this may not be immediately obvious to people who read this code, so the code continues with

```python
assert datamodel is not None  # because we have data
```

First, this tells the reader: don't worry, we're not checking that `datamodel` is not None, because it can't be, and the comment tells why.

Second, if there is a bug in the code, the assertion will fail with `AssertionError`. This doesn't help the user at all -- but the widget would fail later anyway, and failing here makes it easier to find and understand the problem if it occurs.

Gosh, you're getting an advanced course in programming in these comments. :)

Anyway, using assertions is good and we should use them more often.

Reply:
thanks for the in depth explanation, very useful!